optuna/distributions.py

```diff
@@ -509,6 +517,32 @@ def _contains(self, param_value_in_internal_repr: float) -> bool:
         index = int(param_value_in_internal_repr)
         return 0 <= index < len(self.choices)
 
+    def __eq__(self, other: Any) -> bool:
+
+        if not isinstance(other, BaseDistribution):
+            return NotImplemented
+        if type(self) is not type(other):
+            return False
+        if self.__dict__.keys() != other.__dict__.keys():
+            return False
+        for (key, value), (other_key, other_value) in zip(
+            sorted(self.__dict__.items()), sorted(other.__dict__.items())
+        ):
+            if key == "choices":
+                if len(value) != len(other_value):
+                    return False
+                for choice, other_choice in zip(value, other_value):
+                    if not _nanequal(choice, other_choice):
+                        return False
+            else:
+                if value != other_value:
+                    return False
+        return True
+
+    def __hash__(self) -> int:
+
+        return hash((self.__class__,) + tuple(sorted(self.__dict__.items())))
+
```

Comment:
I'm wondering if we can make these lines simple (and reduce the code sizes) by moving the logic of L526-L540 to `check_distribution_compatibility()` function. Do you think there is a benefit that overloading CategoricalDistribution's comparison operators?

Reply:
`CategoricalDistribution`'s comparison operator is called from other codes. So, we need to overload `__eq__` to fix the comparison with NaN.

https://github.com/optuna/optuna/blob/bf8ec607095a4b1d320b1171fa61227e9bf1366a/optuna/distributions.py#L724
https://github.com/optuna/optuna/blob/bf8ec607095a4b1d320b1171fa61227e9bf1366a/optuna/importance/_base.py#L92
https://github.com/optuna/optuna/blob/bf8ec607095a4b1d320b1171fa61227e9bf1366a/optuna/samplers/_search_space/intersection.py#L86
https://github.com/optuna/optuna/blob/bf8ec607095a4b1d320b1171fa61227e9bf1366a/optuna/trial/_trial.py#L647

Reply:
Thank you for your reply! I see. It sounds make sense to overload the operator.